torch/csrc/jit/passes/memory_planning.cpp

```diff
@@ -0,0 +1,159 @@
+
+#include <c10/core/ScalarType.h>
+#include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/memory_planning.h>
+#include <torch/csrc/jit/runtime/static/impl.h>
+
+namespace torch {
+namespace jit {
+
+c10::TensorTypePtr checkIsSupported(const Value& value) {
+  TORCH_CHECK(
+      value.type()->cast<TensorType>(),
+      "out isn't a tensortype ",
+      *value.type());
+  auto ttp = value.type()->expect<TensorType>();
+  TORCH_CHECK(
+      ttp->scalarType().has_value(),
+      "This output was profiled but didn't have a scalar type.")
+  TORCH_CHECK(
+      ttp->sizes().concrete_sizes().has_value(),
+      "This output was profiled but doesn't have sizes.")
+  return ttp;
+}
+
+std::pair<std::vector<int64_t>, std::vector<int64_t>> getSizesStrides(
+    const c10::TensorTypePtr& ttp) {
+  auto sizes = ttp->sizes().concrete_sizes().value();
+  std::vector<int64_t> strides;
+  auto _strides = ttp->strides().concrete_sizes();
+  if (_strides.has_value() && _strides.value()[0] != 0) {
+    strides = _strides.value();
+  } else {
+    strides = at::detail::defaultStrides(sizes);
+  }
+  return std::make_pair(sizes, strides);
+}
+
+size_t computeStorageSize(const c10::TensorTypePtr& ttp) {
+  std::vector<int64_t> sizes, strides;
+  std::tie(sizes, strides) = getSizesStrides(ttp);
+  return at::detail::computeStorageNbytes(
+      sizes, strides, elementSize(ttp->scalarType().value()));
+}
+
+std::pair<size_t, std::map<Value*, c10::TensorTypePtr>>
+computeAlignedStorageSize(std::vector<Node*>& nodes) {
+  // TODO: liveness analysis
+  std::map<Value*, c10::TensorTypePtr> managed_tensors;
+  size_t managed_bytes = 0;
+  for (const auto& node : nodes) {
+    for (const auto& output : node->outputs()) {
+      auto ttp = checkIsSupported(*output);
+      managed_bytes += computeStorageSize(ttp);
+      managed_tensors.insert({output, ttp});
+    }
+  }
+
+  auto aligned_size = MemoryPlanner::compute_aligned_tensor_size(managed_bytes);
+  return std::make_pair(aligned_size, managed_tensors);
+}
+
+DeviceType findDeviceType(std::shared_ptr<Graph>& graph) {
```

Comment:
nitpick: TensorExprs has a similar utility function, maybe we could use either one.

Reply:
is this the one? https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/tensorexpr/kernel.cpp

Reply:
`pickDeviceType` 
https://github.com/pytorch/pytorch/blob/a9b0a921d592b328e7e80a436ef065dadda5f01b/torch/csrc/jit/tensorexpr/kernel.cpp#L160